### DISCUSSION TITLE
There is some way to implement "with" operator without declaring records?

### DISCUSSION THREAD

**Discussion by nathan130200** [dotnet/csharplang#7848] [Q&A] [answered]
I love records, but sometimes I feel that they limit me a lot in specific things. For example I want to implement a 'readonly struct' that stores 3 strings that will make up part of a domain name, so far without problem. Only that struct has to be immutable, unless I could implement the `with` operator individually without necessarily declaring that struct as record...

Why types with `record ' modifier, it implements an equality comparer of its own internally, I can't override that method and I need a specific comparator.

You can even implement it with methods. `.With` but that would not be the goal, imagine each defined type having to push A` With ' method just to copy the fields and modify another, it would be chaos.

My question is exactly this, is there any decent way to use the features of the `with 'operator without declaring the type as 'record'?

Sample:
```cs
readonly struct Email { 
// maybe new keywork "immutable" before struct/class to allow use with operator in custom types that are not records.
  readonly string _localPart, _domainPart;

  public string LocalPart {
    // get => & init =>
  }

  public string DomainPart {
    // get => & init =>
  }
}


Email sourceAddr = "myuser@mywebsite.com"; // myuser = LocalPart, mywebsite.com = DomainPart

var targetAddr = sourceAddr with
{
    LocalPart = "otheruser";
}; 

// otheruser = LocalPart, mywebsite.com = DomainPart

```

**Comment by nathan130200** [comment]
Nevermind, `with` works only in struct and record types, for some weird reason in some days ago i received compile erros when i tried use `with` with readonly structs but now compiles without any error.

**Comment by HaloFour** [comment]
And anonymous types.

It was championed to find a way to generalize a pattern for `with` that could be applied to other types but I don't think there's been any movement on it in a while: #162

**Comment by theunrepentantgeek** [comment, marked as answer]
> Why types with `record ' modifier, it implements an equality comparer of its own internally, I can't override that method and I need a specific comparator.

You can override the way a record implements equality - by explicitly declaring `bool Equals(R? other)`. The compiler fills in everything else for you. 

See [this Stack Overflow answer](https://stackoverflow.com/questions/64326511/custom-equality-check-for-c-sharp-9-records) for more details and links to more information.

**Reply by nathan130200** [reply]
Ahh yes, I tried to declare it before and it kept giving a compilation error saying that the my record type already declared an Equals method, all the time. But now I tested it again and it stopped notifying me of the error, which is why I even requested this discussion. Maybe was outdated cache or something else.

Tested and working


```cs
readonly record struct Pair<F, S>
{
    readonly F _first;
    readonly S _second;

    public Pair(F first, S second)
    {
        _first = first;
        _second = second;
    }

    public F First => _first;
    public S Second => _second;

    public bool Equals(Pair<F, S> other)
    {
        return EqualityComparer<F>.Default.Equals(_first, other._first)
            && EqualityComparer<S>.Default.Equals(_second, other._second);
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(
            _first is null ? 0 : _first.GetHashCode(),
            _second is null ? 0 : _second.GetHashCode()
            );
    }
}
```
